### ISSUE TITLE
Interpolation over derived coordinates

### ISSUE THREAD

**Issue by duncanwp** [SciTools/iris#1759]
None of the interpolation interfaces appear to support interpolation over hybrid (derived) coordinates. However I've been looking at interpolation of a single point on a cube with hybrid height and it seems to be fairly straightforward to first ask for a vertical column using any of the interpolation methods available, and subsequently ask for an (altitude) interpolation over that column. 

Could this be built into the iris interpolation routines to save the overhead of these two seperate calls?

**Comment by pelson** [comment]
> Could this be built into the iris interpolation routines to save the overhead of these two seperate calls? 

Yes - that would be excellent. Though I'm a little uncertain of the difficulties in doing so. :+1:

**Comment by duncanwp** [comment]
This is still something that would be nice to have

**Comment by duncanwp** [comment]
I'd still be interested if we can get it implemented!

**Comment by duncanwp** [comment]
Sure, basically my current approach is to interpolate the regularly gridded dimensions (typically lat/lon/time) to extract column(s) of the hybrid dimension, each of which should now also regular - then interpolate those. 

You can get a feel for that here: https://github.com/cedadev/cis/blob/c207b4e3b47f770da404991eedeeb55422817772/cis/collocation/gridded_interpolation.py#L215. 

It may be that this is the most efficient way of doing it already, but I don't have the capacity to support this code any more and I think it would be of wider use to the community within Iris.

**Comment by trexfeathers** [comment]
Have looked in detail this afternoon. Duncan's vertical column approach seems like the best game in town, and something that we could explore integrating into Iris. Notable differences:

- We would need to pass the calculation to our existing interpolation 'engines' for consistency, probably prevents us simple duplication of Duncan's code.
- It is interesting to note that Iris' `AuxFactory` would be able to immediately calculate the derived coordinate values for the target location(s), without using interpolation. Not sure if this offers value though as interpolation on all dimensions is still needed eventually.

**Comment by trexfeathers** [comment]
From discussion between @trexfeathers and @pp-mo:

- Interpolation is impossible over ANY multi-dimensional coordinate, not just the derived ones, so we're kinda talking about two different problems, both of which need solving.
- ANY approach that uses the hybrid Z for interpolation must assume that the hybrid Z is monotonic at all points, so it's a bit brittle
- We believe that [python-stratify](https://github.com/SciTools/python-stratify) could do more of this under-the-hood - we would not need to write our own code to extract the vertical columns.

So the ideal outcome of this work would not be more code within Iris, but an enhancement to the [User Guide page](https://scitools-iris.readthedocs.io/en/stable/userguide/interpolation_and_regridding.html#interpolation) that demonstrates how to achieve this using Stratify. Obviously we would first need to prove that it works!

**Comment by pp-mo** [comment]
> an enhancement to the [User Guide page](https://scitools-iris.readthedocs.io/en/stable/userguide/interpolation_and_regridding.html#interpolation) that demonstrates how to achieve this using Stratify.

Or see .. #5298

**Comment by trexfeathers** [comment]
Update: we have taken the pragmatic decision that python-stratify gets no further resource for improvements, hence why this has paused for a moment.  However it may still be true that the existing stratify release supports this workflow, and we believe an investigation would take approx 1 day. Could be a useful bit of Iris documentation if we are successful; marking as a candidate for 3.17 release sprints.
